### ISSUE TITLE
Spec compliance

### ISSUE THREAD

**Issue by STRML** [MatAtBread/nodent#68]
Noticed this in a chat with @ljharb about fast-async.

Awaiting a literal is actually allowed as per [spec#2.3.3](https://tc39.github.io/ecmascript-asyncawait/#abstract-ops-async-function-await), which indicates that you should wrap in `Promise.resolve(value)`.

Input:

```js
async function foo() {
  await 3;
}
```

Output:

```js
function foo() {
    return new Promise((function ($return, $error) {
        return (3).then((function ($await_1) {
            return $return();
        }).$asyncbind(this, $error), $error);
    }).$asyncbind(this));
}
```

Expected:

```js
function foo() {
    return new Promise((function ($return, $error) {
        return Promise.resolve(3).then((function ($await_1) {
            return $return();
        }).$asyncbind(this, $error), $error);
    }).$asyncbind(this));
}
```

This is particularly important in code where you might not know what you're actually awaiting, like:

```js
async function tellYouLater(sayWhat) {
  await sayWhat;
}
```

**Comment by ljharb** [comment]
Note that this should apply to every value, not just literals.

**Comment by STRML** [comment]
Right, thanks @ljharb :+1:

**Comment by MattiasBuelens** [comment]
nodent does not wrap every awaited value with `Promise.resolve` by default since this has a non-negligible performance overhead (see the "await non-Promise" section in the [ES7 diffferences](https://github.com/MatAtBread/nodent#gotchas-and-es7-compatibility)). In most use cases, you will only await thenable values, so it's faster to just use `.then` directly rather than first wrapping with `Promise.resolve`.

The documentation explains what you can do if you really need to await a non-thenable value. You can choose to have nodent wrap *all* awaited values with `Promise.resolve` by setting the `wrapAwait` option (and thus taking a performance penalty for all `await`s), or you can manually wrap the awaited values with `Promise.resolve` in your code where you know that you might have a non-thenable value.

**Comment by STRML** [comment]
I apologize for missing that. The conversation we were having was about
using fast-async as a replacement for the standard regenerator-based
transformer, and the not-quite-ready kneden Promise transformer.

In that light, fast-async is comparable, but for users who value spec
compliance and purity (ie no patching prototypes) it appears a few options
are necessary. We'll keep evaluating and thanks for the incredible work.

**Comment by MattiasBuelens** [comment]
@STRML Note: I'm not a contributor to nodent, I'm just a user. I'm also evaluating alternatives for my current build chain setup for `async`/`await`, and nodent is an interesting candidate.

**Comment by STRML** [comment]
@MattiasBuelens Hah, that's what I get for responding by email early in the morning and only skimming your username. Regardless, thanks for the info!

**Comment by ljharb** [comment]
There'd definitely need to be a way to avoid all the caveats and result in fully spec-compliant code, otherwise it's not async/await, it's something else. Are there options for all the things listed in the es7 differences section?

**Comment by STRML** [comment]
Nodent is built as a superset of the spec, and `fast-async` doesn't support any of the new syntax nodent supports, so my answer is a tentative "yes", with the exception of the `case without break` issue listed in the compat section.

**Comment by MatAtBread** [comment]
I'd be prepared to add a "compatibility" mode to fast-async, since the default flags are specified in the Babel plugin, note the nodent codebase. If there are any preferences or suggestions on this, please let me know. 

And happy Christmas :)

**Comment by ljharb** [comment]
@MatAtBread essentially what I and Airbnb needs is a babel plugin that complies with the spec as close to 100% as possible, without a heavyweight runtime dep like `regenerator-runtime`, where any pattern that it doesn't transpile can be prevented via a linter rule, and where performance concerns are always secondary to compliance. If a "compatibility" mode would provide that, that would be amazing, and would enable us to recommend `async`/`await` in https://github.com/airbnb/javascript, as well as use it in our internal codebases.

**Comment by MatAtBread** [comment]
It's reasonably easy to do - nodent implements a super-set of the spec (extensions are in the readme), and the non-Promise await optimizations is disabled via a flag (so you can await on anything). The runtime is pretty small (smaller than regenerator by a considerable amount), and is also optional, although at the cost of littering your code with `try... catch` which can stall the V8 optimizer. If you're prepared to forgo peak performance, simply using the options `promises`, `wrapAwait` and `noRuntime` will do exactly what you want. If you're using `fast-async` you can set these as Babel options (I'd dig out an example, but I'm on my mobile waiting for Santa). 

Are you using Babel, pre-compiling and shipping into multiple runtimes, or into a known version of Node?

Note, this doesn't affect your source or usage of async/await, but might affect the optimal set of options.

The issue about `case` not followed by `break` not transpiling correctly is simply me being lazy. It's fixable, I've just not got around to it.

**Comment by MatAtBread** [comment]
Duh - you're using Babel. Then just use fast-async with the options listed above - it's should all work first time, except the `case` issue

**Comment by ljharb** [comment]
Yes, we're using babel - it would be super helpful if the babel transform had a one-flag option that enabled all the spec compliance options. The core transforms take a `spec: true` option - I'd be very pleased to enable that :-)

**Comment by MatAtBread** [comment]
I'll have a go when I get a free hour or two.

**Comment by matAtWork** [comment]
@STRML @ljharb - if you're feeling brave, install `fast-async` from the branch at https://github.com/MatAtBread/fast-async/tree/6.1.3-spec-option, which implements the `spec` plugin option. Let me know if it does what you need, and I'll publish. 

I'll close this here, as this is actually a `fast-async` issue - please report any problems you find with the `spec` flag in https://github.com/MatAtBread/fast-async/issues

**Comment by STRML** [comment]
Thanks - this is starting to [look really good in combination with the `es6target` option](http://nodent.mailed.me.uk/#async%20function%20tellYouLater(sayWhat)%20%7B%0A%20%20%20%20%2F%2F%20Do%20something%20asynchronous%2C%20such%20as%20DB%20access%2C%20web%20access%2C%20etc.%0A%20%20%20%20const%20result%20%3D%20await%20somethingAsync(sayWhat)%3B%0A%20%20%20%20const%20etwasAnders%20%3D%20await%20somethingElseAsync(sayWhat)%3B%0A%20%20%20%20return%20result%3B%0A%7D%0A~options~%7B%22mode%22%3A%22promises%22%2C%22noRuntime%22%3Atrue%2C%22es6target%22%3Atrue%2C%22promiseType%22%3A%22Zousan%22%7D).

**Comment by matAtWork** [comment]
FYI, I've published `fast-async` v6.2.0 (implements the 'spec' option) which uses the latest nodent (implements the es6target option) and acorn-es7-plugin (fixes an issue with static class methods called 'get'). Hopefully that'll keep people who value compliance over other factors happy.
